File: algorithms/137.single-number-ii.cpp
class Solution {
public:
    // O(n) solution with 32 bits
    int singleNumber(vector<int>& nums) {
        // bit summation & mod 3 operations
        int bits[32] = {0,};
        for (int i=0; i < nums.size(); i++){
            for (int j=0; j < 32; j++){
                int mask = 1<<j;
                int add_bit = (nums[i]&mask) ? 1 : 0;
                bits[j] = (bits[j] + add_bit)%3;
            }
        }
                
        // covert bits -> digial number
        int num = 0;
        for (int j=0; j < 32; j++)
            num += (1 << j)*bits[j];
        return num;
    }
};

File: algorithms/137.single-number-ii.java
public class Solution {
	// O(n) * 32 solution with bitwise solutions
	// Iterate through each bit position (from 0 to 31 for 32-bit integers)
	public int singleNumber(int[] nums) {
		int ans = 0;
		for (int i = 0; i < 32; ++i) {
			int sum = 0;
			for (final int num : nums)
				sum += (num >> i) & 1;  // get the i-th bit
			sum %= 3;           // filter out 3-occurence numbers
			ans |= sum << i;    // Set the i-th bit of the single number
		}

		// Return the final result number which is the single number
		return ans;
	}

	// O(NlogN) solution with sorting solution
	public int singleNumberWithSort(int[] nums) {
		Arrays.sort(nums);

		for(int i=0;i<nums.length-1;i+=3){
			if(nums[i]!=nums[i+1]){
				return nums[i];
			}
			
		}
		 return nums[nums.length-1];
	}
}
